src/basic/fileio.c

```diff
@@ -548,7 +548,9 @@ int read_full_stream_full(
                 }
 
                 buf = t;
-                n = n_next;
+                /* Unless a size has been explicitly specified, try to read as much as fits into the memory
+                 * we allocated (minus 1, to leave one byte for the safety NUL byte) */
+                n = size == SIZE_MAX ? malloc_usable_size(buf) - 1 : n_next;
```

Comment:
not sure I fully grok this - the buffer is allocated asking for n_next which changes on each iteration anyway, what does the value of `size` have to do with it?

Reply:
the function has two modes: it may be called with size == SIZE_MAX, which is the usual case in which case it reads as much as it can (well, subject to the `READ_FULL_BYTES_MAX` safety limit). Or it is called with size != SIZE_MAX, in which case we'll try to read the specified number of bytes, but not more. (this is for example used by systemd-cryptsetup to read keys out of files/partitions based on offset+size). Now, in the former case we have no idea initially how long the buffer is going to be in the end, we start small hence and keep growing it hence. In the latter case however, we know exactly what size to expect, hence we allocate it in full already when starting things, and we'll never use more of it then specified.

Now, in the former case, when growing malloc() sizes we can use `malloc_usable_size()` to make use of the full allocated memory block, and slightly optimize our loop. In the latter case there's no point in that, we wouldn't possibly make use of that anyway.

Reply:
Ok, that makes sense now, thanks